apps/heft/src/plugins/NodeServicePlugin.ts

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as child_process from 'child_process';
+import * as process from 'process';
+import { performance } from 'perf_hooks';
+import { InternalError } from '@rushstack/node-core-library';
+
+import { HeftSession } from '../pluginFramework/HeftSession';
+import { HeftConfiguration } from '../configuration/HeftConfiguration';
+import { IBuildStageContext, ICompileSubstage, IPostBuildSubstage } from '../stages/BuildStage';
+import { ScopedLogger } from '../pluginFramework/logging/ScopedLogger';
+import { IHeftPlugin } from '../pluginFramework/IHeftPlugin';
+import { CoreConfigFiles } from '../utilities/CoreConfigFiles';
+import { SubprocessTerminator } from '../utilities/subprocess/SubprocessTerminator';
+
+const PLUGIN_NAME: string = 'NodeServicePlugin';
+
+export interface INodeServicePluginCompleteConfiguration {
+  commandName: string;
+  ignoreMissingScript: boolean;
+  waitBeforeRestartMs: number;
+  waitForTerminateMs: number;
+  waitForKillMs: number;
+}
+
+export interface INodeServicePluginConfiguration extends Partial<INodeServicePluginCompleteConfiguration> {}
+
+enum State {
+  Stopped,
+  Running,
+  Stopping,
+  Killing
+}
+
+export class NodeServicePlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+  private _logger!: ScopedLogger;
+
+  private _activeChildProcess: child_process.ChildProcess | undefined;
+
+  private _state: State = State.Stopped;
+
+  private _timeout: NodeJS.Timeout | undefined = undefined;
+
+  private _isWindows!: boolean;
+
+  // The process will be automatically restarted when performance.now() exceeds this time
+  private _restartTime: number | undefined = undefined;
+
+  private _configuration!: INodeServicePluginCompleteConfiguration;
+  private _nodeServiceConfiguration: INodeServicePluginConfiguration | undefined = undefined;
+  private _shellCommand: string | undefined;
+
+  // If true, then we will not attempt to relaunch the service until it is rebuilt
+  private _childProcessFailed: boolean = false;
+
+  private _pluginEnabled: boolean = false;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    this._logger = heftSession.requestScopedLogger('node-service');
```

Comment:
I've avoided putting the logger in an instance object in case we eventually call plugins with multiple HeftSession objects in the same Heft process.

Reply:
I feel like `IHeftPlugin` would need to overhaul its design to support that.  I'm not even sure it would be wise.

If separate instances of the same plugin want to share some state, that seems like an advanced scenario that should require a special API, rather than a complexity that we should force every plugin to consider.